lib/spack/spack/compilers/__init__.py

```diff
@@ -203,6 +214,17 @@ def find(compiler_spec, scope=None, init_config=True):
             if c.satisfies(compiler_spec)]
 
 
+@_auto_compiler_spec
+def find_by_arch(compiler_spec, arch_spec, scope=None, init_config=True):
```

Comment:
I'm having trouble distinguishing what this does above and beyond `compilers_for_spec`

Reply:
It returns the CompilerSpec objects, not the Compiler objects.

Reply:
@becker33 

OK, the name seems odd given that it just extracts the specs. I think a more accurate name is `compiler_specs_for_spec` or `available_compilers` or `available_compilers_for_spec` or `available_compiler_specs`.

The `abi` module at least is already working with the `Compiler` objects, so I think it would also be OK not to have a separate function.

Other than that this looks good to me

Reply:
@scheibelp 
I changed it to `find_specs_by_arch`. I looked at the other methods in `spack.compilers`, including `spack.compilers.find` which does the same thing without filtering by arch, and the ideas you had to try to find the best balance for future maintainability.

Let me know if you still have objections with the current name and we can iterate more for clarity.